modules/rostests/apitests/advapi32/Hash.c

```diff
@@ -0,0 +1,298 @@
+       ctx.State[2] == 0xD9DD1BB1 && ctx.State[3] == 0xDEB286C6,
+       "MD5_CTX::State incorrect after MD5Final\n");
+    ok(RtlCompareMemoryUlong(ctx.Buffer, sizeof(ctx.Buffer), 0) == sizeof(ctx.Buffer),
+       "MD5_CTX::Buffer incorrect after MD5Final\n");
+    ok(RtlCompareMemory(ctx.Hash, &g_aucMD5Result, MD5_DIGEST_LEN) == MD5_DIGEST_LEN,
+       "MD5 result incorrect\n");
+}
+
+static void Test_MD4()
+{
+    FN_MD4Init* pfnMD4Init;
+    FN_MD4Update* pfnMD4Update;
+    FN_MD4Final* pfnMD4Final;
+    DECLSPEC_ALIGN(4) MD4_CTX ctx;
+
+    /* Load functions */
+    pfnMD4Init = (FN_MD4Init*)GetProcAddress(g_hAdvapi32, "MD4Init");
+    pfnMD4Update = (FN_MD4Update*)GetProcAddress(g_hAdvapi32, "MD4Update");
+    pfnMD4Final = (FN_MD4Final*)GetProcAddress(g_hAdvapi32, "MD4Final");
+
+    if (pfnMD4Init == NULL || pfnMD4Update == NULL || pfnMD4Final == NULL)
+    {
+        skip("advapi32.dll!MD4XXX not found\n");
+        return;
+    }
+    
+    /* Test MD4Init */
+    pfnMD4Init(&ctx);
+    ok(ctx.Count[0] == 0 && ctx.Count[1] == 0, "MD4_CTX::Count incorrect after MD4Init\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_aulMD5Or4StateInit,
+                        RTL_FIELD_SIZE(MD4_CTX, State)) == RTL_FIELD_SIZE(MD4_CTX, State),
+       "MD4_CTX::State incorrect after MD4Init\n");
+
+    /* Test MD4Update */
+    pfnMD4Update(&ctx, (PUCHAR)g_TestString.Buffer, g_TestString.Length);
+    ok(RtlCompareMemory(ctx.Buffer,
+                        g_TestString.Buffer,
+                        g_TestString.Length) == g_TestString.Length,
+       "MD4_CTX::Buffer incorrect after MD4Update\n");
+    ok(RtlCompareMemory(ctx.State,
+                        g_aulMD5Or4StateInit,
+                        RTL_FIELD_SIZE(MD4_CTX, State) == RTL_FIELD_SIZE(MD4_CTX, State)),
+       "MD4_CTX::State incorrect after MD4Update\n");
+    ok(ctx.Count[0] == g_TestString.Length * CHAR_BIT && ctx.Count[1] == 0,
+       "MD4_CTX::Count incorrect after MD4Update\n");
+
+    /* Test MD4Final */
+    pfnMD4Final(&ctx);
+    ok(ctx.Count[0] == 0x200 && ctx.Count[1] == 0,
+       "MD4_CTX::Count [0x%08lX, 0x%08lX] is incorrect after MD4Final, expect [0x200, 0]\n",
+       ctx.Count[0],
+       ctx.Count[1]);
+    ok(ctx.State[0] == 0x8A50E8E0 && ctx.State[1] == 0xA602114D &&
+       ctx.State[2] == 0xABA7F06A && ctx.State[3] == 0xED40C4D8,
+       "MD4_CTX::State incorrect after MD4Final\n");
+    ok(RtlCompareMemoryUlong(ctx.Buffer, sizeof(ctx.Buffer), 0) == sizeof(ctx.Buffer),
+       "MD4_CTX::Buffer incorrect after MD4Final\n");
+    ok(RtlCompareMemory(ctx.Hash, &g_aucMD4Result, MD4_DIGEST_LEN) == MD4_DIGEST_LEN,
+       "MD4 result incorrect\n");
+}
+
+START_TEST(Hash)
+{
+    /* Load advapi32.dll */
+    g_hAdvapi32 = GetModuleHandleW(L"advapi32.dll");
+    if (!g_hAdvapi32)
+    {
+        skip("Module advapi32 not found\n");
+        return;
+    }
+
+    Test_SHA1();
+    Test_MD5();
+    Test_MD4();
```

Comment:
From oldest to newest...
```suggestion
    Test_MD4();
    Test_MD5();
    Test_SHA1();
```

Reply:
Why?

Reply:
Personal suggestion: timeline/alphabetical order,